# Import needed libraries

In [1]:
import pandas as pd # data processing, CSV file I/O 
import numpy as np # linear algebra
import matplotlib.pyplot as plt # create graphs
from dateutil import parser #parsing dates
import seaborn as sns # create graphs

# Load data
data_link: https://www.kaggle.com/datasets/volpatto/coffee-quality-database-from-cqi?select=merged_data_cleaned.csv

In [2]:
# Read the CSV file into a pandas DataFrame and store it in the variable 'df'.
# The file path provided should point to the location of the CSV file on your system.
df = pd.read_csv("merged_data_cleaned.csv")

# Display the first few rows of the DataFrame to quickly inspect the data.
# It helps to get a quick overview of the data structure and contents.
df.head()


Unnamed: 0  Species                     Owner Country.of.Origin  \
0           0  Arabica                 metad plc          Ethiopia   
1           1  Arabica                 metad plc          Ethiopia   
2           2  Arabica  grounds for health admin         Guatemala   
3           3  Arabica       yidnekachew dabessa          Ethiopia   
4           4  Arabica                 metad plc          Ethiopia   

                                  Farm.Name Lot.Number       Mill ICO.Number  \
0                                 metad plc        NaN  metad plc  2014/2015   
1                                 metad plc        NaN  metad plc  2014/2015   
2  san marcos barrancas "san cristobal cuch        NaN        NaN        NaN   
3     yidnekachew dabessa coffee plantation        NaN    wolensu        NaN   
4                                 metad plc        NaN  metad plc  2014/2015   

                                 Company       Altitude  ...  Color  \
0      metad agricultural developmet plc      1950-2200  ...  Green   
1      metad agricultural developmet plc      1950-2200  ...  Green   
2                                    NaN  1600 - 1800 m  ...    NaN   
3  yidnekachew debessa coffee plantation      1800-2200  ...  Green   
4      metad agricultural developmet plc      1950-2200  ...  Green   

  Category.Two.Defects        Expiration                  Certification.Body  \
0                    0   April 3rd, 2016  METAD Agricultural Development plc   
1                    1   April 3rd, 2016  METAD Agricultural Development plc   
2                    0    May 31st, 2011        Specialty Coffee Association   
3                    2  March 25th, 2016  METAD Agricultural Development plc   
4                    2   April 3rd, 2016  METAD Agricultural Development plc   

                      Certification.Address  \
0  309fcf77415a3661ae83e027f7e5f05dad786e44   
1  309fcf77415a3661ae83e027f7e5f05dad786e44   
2  36d0d00a3724338ba7937c52a378d085f2172daa   
3  309fcf77415a3661ae83e027f7e5f05dad786e44   
4  309fcf77415a3661ae83e027f7e5f05dad786e44   

                      Certification.Contact unit_of_measurement  \
0  19fef5a731de2db57d16da10287413f5f99bc2dd                   m   
1  19fef5a731de2db57d16da10287413f5f99bc2dd                   m   
2  0878a7d4b9d35ddbf0fe2ce69a2062cceb45a660                   m   
3  19fef5a731de2db57d16da10287413f5f99bc2dd                   m   
4  19fef5a731de2db57d16da10287413f5f99bc2dd                   m   

  altitude_low_meters altitude_high_meters altitude_mean_meters  
0              1950.0               2200.0               2075.0  
1              1950.0               2200.0               2075.0  
2              1600.0               1800.0               1700.0  
3              1800.0               2200.0               2000.0  
4              1950.0               2200.0               2075.0  

[5 rows x 44 columns]

# Basic exploration

In [3]:
# Retrieve the dimensions of the DataFrame (number of rows, number of columns).
shape = df.shape

# The first element of the tuple shape represents the number of rows, while the second element represents the number of columns.
# Print the dimensions of the DataFrame.
print("Number of rows:", shape[0])
print("Number of columns:", shape[1])


Number of rows: 1339
Number of columns: 44


In [4]:
# Retrieve the column names (labels) of the DataFrame 'df'.
columns = df.columns

# Print the column names of the DataFrame.
print(columns)

Index(['Unnamed: 0', 'Species', 'Owner', 'Country.of.Origin', 'Farm.Name',
       'Lot.Number', 'Mill', 'ICO.Number', 'Company', 'Altitude', 'Region',
       'Producer', 'Number.of.Bags', 'Bag.Weight', 'In.Country.Partner',
       'Harvest.Year', 'Grading.Date', 'Owner.1', 'Variety',
       'Processing.Method', 'Aroma', 'Flavor', 'Aftertaste', 'Acidity', 'Body',
       'Balance', 'Uniformity', 'Clean.Cup', 'Sweetness', 'Cupper.Points',
       'Total.Cup.Points', 'Moisture', 'Category.One.Defects', 'Quakers',
       'Color', 'Category.Two.Defects', 'Expiration', 'Certification.Body',
       'Certification.Address', 'Certification.Contact', 'unit_of_measurement',
       'altitude_low_meters', 'altitude_high_meters', 'altitude_mean_meters'],
      dtype='object')


In [5]:
# Display concise summary information about the DataFrame 'df'.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1339 entries, 0 to 1338
Data columns (total 44 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             1339 non-null   int64  
 1   Species                1339 non-null   object 
 2   Owner                  1332 non-null   object 
 3   Country.of.Origin      1338 non-null   object 
 4   Farm.Name              980 non-null    object 
 5   Lot.Number             276 non-null    object 
 6   Mill                   1021 non-null   object 
 7   ICO.Number             1182 non-null   object 
 8   Company                1130 non-null   object 
 9   Altitude               1113 non-null   object 
 10  Region                 1280 non-null   object 
 11  Producer               1107 non-null   object 
 12  Number.of.Bags         1339 non-null   int64  
 13  Bag.Weight             1339 non-null   object 
 14  In.Country.Partner     1339 non-null   object 
 15  Harv

In [6]:
# Calculate the number of missing values for each column using the 'isnull()' method.
missing_values_count = df.isnull().sum()

# Calculate the percentage of missing values for each column by dividing the number of missing values by the total number of rows in the DataFrame.
# missing_values_count is multiplied by 100 to convert it to a percentage.
percent_missing = missing_values_count * 100 / len(df)

# Display the percentage of missing values for each column.
print(percent_missing)

Unnamed: 0                0.000000
Species                   0.000000
Owner                     0.522778
Country.of.Origin         0.074683
Farm.Name                26.811053
Lot.Number               79.387603
Mill                     23.749066
ICO.Number               11.725168
Company                  15.608663
Altitude                 16.878267
Region                    4.406273
Producer                 17.326363
Number.of.Bags            0.000000
Bag.Weight                0.000000
In.Country.Partner        0.000000
Harvest.Year              3.510082
Grading.Date              0.000000
Owner.1                   0.522778
Variety                  16.878267
Processing.Method        12.696042
Aroma                     0.000000
Flavor                    0.000000
Aftertaste                0.000000
Acidity                   0.000000
Body                      0.000000
Balance                   0.000000
Uniformity                0.000000
Clean.Cup                 0.000000
Sweetness           

In [7]:
# Access the 'Species' column of the DataFrame 'df' and retrieve the counts of each unique value to see categories of coffee Speciesin this dataset.
# The 'value_counts()' method counts the occurrences of each unique value in the 'Species' column and returns a Series.
species_counts = df['Species'].value_counts()

# Print the counts of each unique value in the 'Species' column, displaying them in descending order.
print(species_counts)


Arabica    1311
Robusta      28
Name: Species, dtype: int64


# Understanding the data

The Arabica coffee data contains 44 columns and 1339 rows.

Of the 44 columns, we are only interested in some of these columns which are Unnamed(representing id), Country.of.Origin, Producer, Harvest.Year, Grading.Date, Processing.Method, Aroma, Flavor, Aftertaste, Acidity, Body, Balance, Cupper.Points, Total.Cup.Points, Moisture, Category.One.Defects, Color, Category.Two.Defects, Expiration.

Since the dataset is relatively small with few rows, it's crucial to handle any null values appropriately to maximize data utilization.

This dataset consists of only two species (Arabica and Robusta), with one category having a significantly lower number of rows compared to the other.We'll focus on analyzing Arabica coffee samples with a higher number of rows to ensure reliable insights and statistical significance.

In [8]:
# Create a new DataFrame 'df2' by dropping certain columns that are not needed for the analysis.
# Columns such as 'Lot.Number', 'Mill', 'ICO.Number', etc., are dropped to focus only on relevant data for the analysis.
df2 = df.drop(['Producer','Owner.1','Region','Owner','Company','Lot.Number', 'Mill', 'ICO.Number', 'Number.of.Bags', 'Bag.Weight', 'Variety', 'Farm.Name', 'Certification.Address', 'Certification.Contact', 'Certification.Body', 'In.Country.Partner', 'altitude_mean_meters', 'altitude_high_meters', 'altitude_low_meters', 'Species', 'Sweetness', 'Quakers', 'Clean.Cup', 'Uniformity'], axis=1)

In [9]:
# Create a new column 'Defects' in DataFrame 'df2' by summing up the values from 'Category.One.Defects' and 'Category.Two.Defects'.
# This new column represents the total number of defects for each entry, combining defects from two different categories.
df2['Defects'] = df2['Category.One.Defects'] + df2['Category.Two.Defects']

In [10]:
# Drop the columns 'Category.One.Defects' and 'Category.Two.Defects' from DataFrame 'df2'.
# These columns are dropped as they have been combined into a new column 'Defects'.
df2.drop(['Category.One.Defects', 'Category.Two.Defects'], axis=1, inplace=True)

In [11]:
# Calculate the number of missing values for each column using the 'isnull()' method.
missing_values_count = df2.isnull().sum()

# Calculate the percentage of missing values for each column by dividing the number of missing values by the total number of rows in the DataFrame.
# missing_values_count is multiplied by 100 to convert it to a percentage.
percent_missing = missing_values_count * 100 / len(df2)

# Display the percentage of missing values for each column.
print(percent_missing)

Unnamed: 0              0.000000
Country.of.Origin       0.074683
Altitude               16.878267
Harvest.Year            3.510082
Grading.Date            0.000000
Processing.Method      12.696042
Aroma                   0.000000
Flavor                  0.000000
Aftertaste              0.000000
Acidity                 0.000000
Body                    0.000000
Balance                 0.000000
Cupper.Points           0.000000
Total.Cup.Points        0.000000
Moisture                0.000000
Color                  16.280807
Expiration              0.000000
unit_of_measurement     0.000000
Defects                 0.000000
dtype: float64


# Basic Preprocessing

In [12]:
# Print rows where the 'Country.of.Origin' column is missing (null)
# This helps to identify entries where the country of origin is not specified
print(df2[pd.isnull(df2['Country.of.Origin'])])

      Unnamed: 0 Country.of.Origin Altitude Harvest.Year     Grading.Date  \
1197        1197               NaN      NaN          NaN  March 1st, 2011   

     Processing.Method  Aroma  Flavor  Aftertaste  Acidity  Body  Balance  \
1197               NaN   6.75    6.75        6.42     6.83  7.58      7.5   

      Cupper.Points  Total.Cup.Points  Moisture Color           Expiration  \
1197           7.25             79.08       0.1   NaN  February 29th, 2012   

     unit_of_measurement  Defects  
1197                   m        3  


In [13]:
# Fill missing values in the 'Country.of.Origin' column with 'Colombia' based on the owner name
# This assumes that if the country of origin is missing, it's should be as the owner's nationality
df2['Country.of.Origin'].fillna('Colombia', inplace=True)

# Normalize the 'Country.of.Origin' column to lower case
df2['Country.of.Origin'] = df['Country.of.Origin'].str.lower()

In [14]:
# Mapping the Country
Country_of_Origin = {
    "united states (hawaii)": "hawaii, united states",
    "united states (puerto rico)": "puerto rico, united states",
    "cote d?ivoire": "cote divoire"
}

# Define the fixCountry function
def fixCountry(c):
    if c == 'united states (hawaii)' or c == 'united states (puerto rico)':
        return "united states"
    elif c == "cote d?ivoire":
        return 'cote divoire'
    else:
        return c

# Applying the function to fix the values in the column
df2['Country.of.Origin'] = df2['Country.of.Origin'].apply(fixCountry)

In [15]:
# Define a function to clean altitude range values and calculate the mean altitude
def clean_altitude_range(range_value):
    if isinstance(range_value, str):
        range_value = range_value.replace(" ", "")  # Remove blank spaces
        if '-' in range_value:
            try:
                start, end = range_value.split('-')
                start = int(start)
                end = int(end)
                return (start + end) / 2  # Calculate the mean altitude for the range
            except ValueError:
                return np.nan  # Return NaN if the range cannot be parsed
        else:
            try:
                return int(range_value)  # Convert single altitude value to int
            except ValueError:
                return np.nan  # Return NaN if the value cannot be parsed
    else:
        return range_value  # Return the original value if it's not a string


# Apply the clean_altitude_range function to clean and calculate the mean altitude for each value in the "Altitude" column
df2['Altitude'] = df2['Altitude'].apply(clean_altitude_range)

# Fill missing values in the 'Altitude' column with the mean altitude
df2['Altitude'].fillna(df2['Altitude'].mean(), inplace=True)

In [16]:
# Drop rows with null values in the 'Harvest.Year' column that cannot be reasonably estimated
df2.dropna(subset=['Harvest.Year'], inplace=True)

In [17]:
# Fill missing values in the 'Processing.Method' column with the most frequent value
df2['Processing.Method'].fillna(df2['Processing.Method'].mode().iloc[0], inplace=True)

In [18]:
# Fill missing values in the 'Color' column with the most frequent value
df2['Color'].fillna(df2['Color'].mode().iloc[0], inplace=True)

In [19]:
# Set 'Harvest.Year' with the first year and remove any spaces
df2['Harvest.Year'] = df2['Harvest.Year'].str.split('/').str[0].str.strip().str[:4]

# Clean the 'Harvest.Year' column by removing any non-numeric characters
df2['Harvest.Year'] = df2['Harvest.Year'].str.replace(r'[^0-9]', '')


C:\Users\Abdo Mostafa\AppData\Local\Temp\ipykernel_1900\3659457714.py:5: FutureWarning: The default value of regex will change from True to False in a future version.
  df2['Harvest.Year'] = df2['Harvest.Year'].str.replace(r'[^0-9]', '')


In [20]:
# More values cleaning
harvest_year = list(df2['Harvest.Year'])
new = []

# Iterate through each element in the 'Harvest.Year' list
for i in range(len(harvest_year)):
    # Check if the length of the 'Harvest.Year' string at index i is less than 4
    if len(harvest_year[i]) < 4:
        # If the year has less than 4 digits, add "20" to the beginning of the year
        new.append(str(2 * (10 ** (3 - len(harvest_year[i])))) + harvest_year[i])
    else:
        # If the year has 4 or more digits, keep it unchanged
        new.append(harvest_year[i])

# Update the 'Harvest.Year' column with the cleaned values
df2['Harvest.Year'] = new

In [21]:
# Convert 'Harvest.Year' and 'Expiration' columns to datetime objects using dateutil parser
df2['Harvest.Year'] = df2['Harvest.Year'].apply(parser.parse)
df2['Expiration'] = df2['Expiration'].apply(parser.parse)

In [22]:
# # Subtracts the "Harvest Year" from the "Expiration" date and accesses the 'days' attribute of the resulting timedelta object
# df2['Coffee Age'] = (df2['Expiration'] - df2['Harvest.Year']).dt.days

In [23]:
# # Retrieve the indices of rows where the 'Coffee Age' is less than 0
# # This filters the DataFrame to find rows where the difference between 'Expiration' and 'Harvest Year' is negative
# index_of_negative_age = df2[df2['Coffee Age'] < 0].index

# # Print or use the indices as needed
# print(index_of_negative_age)

In [24]:
# # Drop the rows where the 'Coffee Age' is less than 0 using their indices
# df2.drop(df2[df2['Coffee Age'] < 0].index, inplace=True)

In [25]:
df2['Grading.Date'].unique()

array(['April 4th, 2015', 'March 26th, 2015', 'September 3rd, 2013',
       'September 17th, 2012', 'September 2nd, 2010', 'March 30th, 2015',
       'March 27th, 2015', 'March 13th, 2015', 'May 31st, 2010',
       'August 31st, 2010', 'June 16th, 2010', 'April 7th, 2016',
       'March 25th, 2015', 'April 2nd, 2014', 'July 26th, 2012',
       'December 3rd, 2011', 'June 26th, 2014', 'May 18th, 2017',
       'May 18th, 2016', 'December 3rd, 2015', 'May 30th, 2014',
       'April 6th, 2012', 'January 28th, 2011', 'June 17th, 2010',
       'April 12th, 2012', 'January 4th, 2011', 'June 10th, 2015',
       'January 29th, 2015', 'October 1st, 2014', 'March 20th, 2014',
       'February 6th, 2014', 'February 13th, 2012', 'May 16th, 2016',
       'June 30th, 2014', 'July 26th, 2013', 'October 9th, 2017',
       'July 3rd, 2012', 'June 5th, 2012', 'June 23rd, 2017',
       'January 22nd, 2014', 'October 25th, 2012', 'March 18th, 2013',
       'January 12th, 2013', 'May 23rd, 2015', 'March 31s

In [27]:
from dateutil import parser

df2['Grading.Date']=df2['Grading.Date'].apply(parser.parse)


In [28]:
from dateutil import parser

# Assuming df2 is your DataFrame with the column 'Grading.Date'
#df2['Grading.Date'] = df2['Grading.Date'].apply(parser.parse)


# Print the DataFrame to verify the conversion
print(df2)


      Unnamed: 0 Country.of.Origin     Altitude         Harvest.Year  \
0              0          ethiopia  2075.000000  2014-05-04 00:00:00   
1              1          ethiopia  2075.000000  2014-05-04 00:00:00   
3              3          ethiopia  2000.000000  2014-05-04 00:00:00   
4              4          ethiopia  2075.000000  2014-05-04 00:00:00   
5              5            brazil  1855.490483  2013-05-04 00:00:00   
...          ...               ...          ...                  ...   
1334        1334           ecuador  1855.490483  2016-05-04 00:00:00   
1335        1335           ecuador    40.000000  2016-05-04 00:00:00   
1336        1336     united states  1855.490483  2014-05-04 00:00:00   
1337        1337             india  1855.490483  2013-05-04 00:00:00   
1338        1338           vietnam  1855.490483  2013-05-04 00:00:00   

     Grading.Date Processing.Method  Aroma  Flavor  Aftertaste  Acidity  Body  \
0      2015-04-04      Washed / Wet   8.67    8.83    

In [29]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1292 entries, 0 to 1338
Data columns (total 19 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   Unnamed: 0           1292 non-null   int64         
 1   Country.of.Origin    1292 non-null   object        
 2   Altitude             1292 non-null   float64       
 3   Harvest.Year         1292 non-null   object        
 4   Grading.Date         1292 non-null   datetime64[ns]
 5   Processing.Method    1292 non-null   object        
 6   Aroma                1292 non-null   float64       
 7   Flavor               1292 non-null   float64       
 8   Aftertaste           1292 non-null   float64       
 9   Acidity              1292 non-null   float64       
 10  Body                 1292 non-null   float64       
 11  Balance              1292 non-null   float64       
 12  Cupper.Points        1292 non-null   float64       
 13  Total.Cup.Points     1292 non-nul

In [40]:
df2['Harvest.Year']

0       2014-05-04 00:00:00
1       2014-05-04 00:00:00
3       2014-05-04 00:00:00
4       2014-05-04 00:00:00
5       2013-05-04 00:00:00
               ...         
1334    2016-05-04 00:00:00
1335    2016-05-04 00:00:00
1336    2014-05-04 00:00:00
1337    2013-05-04 00:00:00
1338    2013-05-04 00:00:00
Name: Harvest.Year, Length: 1292, dtype: object

In [50]:
# Convert the column to datetime type with error handling
df2['Harvest.Year'] = pd.to_datetime(df2['Harvest.Year'], errors='coerce')

# Print rows with out-of-bounds dates
df2[df2['Harvest.Year'].isnull()]

# Now, you can proceed with the rest of your analysis, keeping in mind that some dates may be missing or converted to NaT (Not a Time).


Empty DataFrame
Columns: [Unnamed: 0, Country.of.Origin, Altitude, Harvest.Year, Grading.Date, Processing.Method, Aroma, Flavor, Aftertaste, Acidity, Body, Balance, Cupper.Points, Total.Cup.Points, Moisture, Color, Expiration, unit_of_measurement, Defects]
Index: []

In [62]:
max(df2['Harvest.Year'].dt.year)

2018

In [59]:
df2[df2['Harvest.Year'].dt.year>2018].index

Int64Index([374, 493], dtype='int64')

In [61]:
df2.drop(df2[df2['Harvest.Year'].dt.year>2018].index,inplace=True)

In [49]:
df2.dropna(inplace=True)

In [46]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1292 entries, 0 to 1338
Data columns (total 19 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   Unnamed: 0           1292 non-null   int64         
 1   Country.of.Origin    1292 non-null   object        
 2   Altitude             1292 non-null   float64       
 3   Harvest.Year         1291 non-null   datetime64[ns]
 4   Grading.Date         1292 non-null   datetime64[ns]
 5   Processing.Method    1292 non-null   object        
 6   Aroma                1292 non-null   float64       
 7   Flavor               1292 non-null   float64       
 8   Aftertaste           1292 non-null   float64       
 9   Acidity              1292 non-null   float64       
 10  Body                 1292 non-null   float64       
 11  Balance              1292 non-null   float64       
 12  Cupper.Points        1292 non-null   float64       
 13  Total.Cup.Points     1292 non-nul

In [63]:
len(list(df2['Grading.Date']))

1289

In [64]:
# Save the DataFrame to a new CSV file
df2.to_csv("coffee_CleanNew_2.csv", index=False)